### PR TITLE
Move AutoRefreshingToken, TokenEndpoint, and TokenRequest

### PR DESCRIPTION
**Pull request by dudnyk**
Move AutoRefreshingToken, TokenEndpoint, and TokenRequest
files to src folder.

Remove deprecated warnings

Move kDefaultMinimumValidity, kDefaultMinimumValiditySeconds, and
kForceRefresh constants from AutoRefreshingToken to TokenProvider

Move internal implementation of TokenProvider(TokenProviderImpl class) to
"internal" namespace.

Relates-To: OLPEDGE-1880

Signed-off-by: Yevhenii Dudnyk <ext-yevhenii.dudnyk@here.com>

### CODE REVIEW

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h [hunk 1]
```diff
@@ -25,22 +25,51 @@
...
+  int GetHttpStatusCode() const;
+
+  TokenResponse GetResponse(client::CancellationContext& context) const;
+
+  bool IsTokenResponseOK() const;
+
+ private:
+  std::chrono::seconds minimum_validity_{kDefaultMinimumValidity};
+  std::shared_ptr<AutoRefreshingToken> token_;
+  mutable std::mutex request_mutex_;
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:70]: Instead of this, I would add another layer to hide even the AutoRefreshingToken, i.e. TokenProviderPrivate.

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h [hunk 2]
```diff
@@ -25,22 +25,51 @@
...
+
+  int GetHttpStatusCode() const;
+
+  TokenResponse GetResponse(client::CancellationContext& context) const;
+
+  bool IsTokenResponseOK() const;
+
+ private:
+  std::chrono::seconds minimum_validity_{kDefaultMinimumValidity};
+  std::shared_ptr<AutoRefreshingToken> token_;
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:69]: AutoRefreshingToken is a pimpl to shared_ptr so once we can hide it lets remove the shared_ptr wrapper.

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h [hunk 3]
```diff
@@ -25,22 +25,51 @@
...
-// Needed to avoid endless warnings from TokenRequest/TokenResult
-PORTING_PUSH_WARNINGS()
-PORTING_CLANG_GCC_DISABLE_WARNING("-Wdeprecated-declarations")
+static constexpr auto kDefaultMinimumValidity = 300ll;
+static constexpr auto kDefaultMinimumValiditySeconds =
+    std::chrono::seconds(kDefaultMinimumValidity);
+static constexpr auto kForceRefresh = std::chrono::seconds(0);
+
+template <uint64_t MinimumValidity>
+class TokenProvider;
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:44]: Needed?

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp [hunk 4]
```diff
@@ -17,15 +17,17 @@
...
 
 #include <chrono>
 #include <iostream>
 
-#include "olp/authentication/TokenEndpoint.h"
+#include "TokenEndpoint.h"
 #include "olp/core/client/CancellationToken.h"
 #include "olp/core/logging/Log.h"
-#include "olp/core/porting/warning_disable.h"
+
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp:28]: Remove, all project headers are grouped together, see https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.h [hunk 5]
```diff
@@ -23,36 +23,24 @@
...
-#include "AuthenticationApi.h"
-#include "TokenEndpoint.h"
-#include "TokenRequest.h"
-#include "TokenResult.h"
+#include <olp/authentication/AuthenticationApi.h>
 
 #include "olp/core/client/CancellationContext.h"
 #include "olp/core/client/CancellationToken.h"
-#include "olp/core/porting/warning_disable.h"
+
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/AutoRefreshingToken.h:30]: Same here.

File: olp-cpp-sdk-authentication/src/TokenEndpoint.h [hunk 6]
```diff
@@ -26,30 +26,25 @@
...
 #include <olp/authentication/Settings.h>
-#include <olp/authentication/TokenRequest.h>
-#include <olp/authentication/TokenResult.h>
 #include <olp/authentication/Types.h>
 #include <olp/core/client/ApiResponse.h>
 #include <olp/core/client/CancellationContext.h>
 #include <olp/core/client/CancellationToken.h>
-#include <olp/core/porting/warning_disable.h>
+
+#include "TokenRequest.h"
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/TokenEndpoint.h:34]: Same here.

File: olp-cpp-sdk-authentication/src/TokenEndpoint.h [hunk 7]
```diff
@@ -60,8 +55,8 @@ class AUTHENTICATION_API OLP_SDK_DEPRECATED("Will be removed in 10.2020")
    * @brief Executes the POST request method to the token endpoint.
    *
    * The request gets the HERE Access token that is used to access the HERE
-   * platform Services. Returns the token that is used as the `Authorization:
-   * Bearer` token value.
+   * platform Services. Returns the token that is used as
+   * the`Authorization: Bearer` token value.
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/TokenEndpoint.h:59]:
```suggestion
   * the `Authorization: Bearer` token value.
```

File: olp-cpp-sdk-authentication/src/TokenEndpointImpl.cpp [hunk 8]
```diff
@@ -18,6 +18,7 @@
  */
 
 #include "TokenEndpointImpl.h"
+#include "TokenResult.h"
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/TokenEndpointImpl.cpp:21]: Move to below group.

File: olp-cpp-sdk-authentication/src/TokenEndpointImpl.h [hunk 9]
```diff
@@ -24,11 +24,11 @@
 #include <olp/authentication/AuthenticationSettings.h>
 #include <olp/authentication/Settings.h>
 #include <olp/authentication/SignInResult.h>
-#include <olp/authentication/TokenRequest.h>
 #include <olp/authentication/Types.h>
 #include <olp/core/client/CancellationContext.h>
 #include <olp/core/client/OlpClient.h>
-#include <olp/core/porting/warning_disable.h>
+
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/TokenEndpointImpl.h:30]: Same

File: olp-cpp-sdk-authentication/src/TokenProvider.cpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+ * License-Filename: LICENSE
+ */
+
+#include <olp/authentication/TokenProvider.h>
+
+#include "AutoRefreshingToken.h"
+#include "TokenEndpoint.h"
+#include "TokenRequest.h"
+#include "TokenResult.h"
+
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/TokenProvider.cpp:26]: Namespace is missing. Nothing should be placed in the global namespace.

File: tests/integration/olp-cpp-sdk-authentication/HereAccountOauth2Test.cpp [hunk 11]
```diff
@@ -23,13 +23,15 @@
 #include <matchers/NetworkUrlMatchers.h>
 #include <mocks/NetworkMock.h>
 #include <olp/authentication/AuthenticationClient.h>
-#include <olp/authentication/AutoRefreshingToken.h>
+#include <olp/authentication/TokenProvider.h>
 #include <olp/core/client/OlpClientSettingsFactory.h>
 #include <olp/core/http/HttpStatusCode.h>
 #include <olp/core/porting/make_unique.h>
 #include <olp/core/porting/warning_disable.h>
```
Comment (andescu) [inline tests/integration/olp-cpp-sdk-authentication/HereAccountOauth2Test.cpp:31]: Remove

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 3 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:44]: yes, cause I use friend for TokenProvider from another namespace(internal)

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h [hunk 12]
```diff
@@ -87,23 +117,6 @@ class TokenProvider {
...
-   * string otherwise.
-   *
-   * @deprecated Will be removed by 10.2022. Use the operator with
-   * `CancellationContext` instead.
-   */
-  OLP_SDK_DEPRECATED(
-      "Will be removed by 10.2022. Use the operator with `CancellationContext` "
-      "instead.")
-  std::string operator()() const { return impl_->operator()(); }
-
```
[the remark this answers is not in the document]
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:106]: ok

File: olp-cpp-sdk-authentication/src/TokenEndpoint.h (hunk 7 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenEndpoint.h:59]: done

File: olp-cpp-sdk-authentication/src/TokenEndpointImpl.cpp (hunk 8 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenEndpointImpl.cpp:21]: done

File: tests/integration/olp-cpp-sdk-authentication/HereAccountOauth2Test.cpp (hunk 11 above)
Comment (dudnyk) [reply tests/integration/olp-cpp-sdk-authentication/HereAccountOauth2Test.cpp:31]: I moved these constants to TokenProvider:
static constexpr auto kDefaultMinimumValidity = 300ll;
static constexpr auto kDefaultMinimumValiditySeconds =
    std::chrono::seconds(kDefaultMinimumValidity);
static constexpr auto kForceRefresh = std::chrono::seconds(0);

File: olp-cpp-sdk-authentication/src/TokenProvider.cpp (hunk 10 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenProvider.cpp:26]: these files are in the same directory.

File: olp-cpp-sdk-authentication/src/TokenEndpointImpl.h (hunk 9 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenEndpointImpl.h:30]: done

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp (hunk 4 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp:28]: done

File: olp-cpp-sdk-authentication/src/TokenEndpoint.h (hunk 6 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenEndpoint.h:34]: done

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 3 above)
Comment (andescu) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:44]: Why do you need a friend class? I think from our point of view we marked that this class is internal, and it is not bound to the API stability and users should not use it directly, so every method can be public. If any user thinks he requires using this class alone, it is his choice.

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp (hunk 4 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp:28]: done

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 2 above)
Comment (Smartfanat) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:69]: done

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 1 above)
Comment (Smartfanat) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:70]: done

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 3 above)
Comment (Smartfanat) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:44]: ok)

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.h (hunk 5 above)
Comment (Smartfanat) [reply olp-cpp-sdk-authentication/src/AutoRefreshingToken.h:31]: done

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp [hunk 13]
```diff
@@ -17,15 +17,16 @@
...
 
 #include <chrono>
 #include <iostream>
 
-#include "olp/authentication/TokenEndpoint.h"
+#include "TokenEndpoint.h"
+#include "TokenRequest.h"
+#include "TokenResult.h"
 #include "olp/core/client/CancellationToken.h"
 #include "olp/core/logging/Log.h"
```
Comment (mykolamalik) [inline olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp:29]:
```suggestion
#include <olp/core/client/CancellationToken.h>
#include <olp/core/logging/Log.h>
#include "TokenEndpoint.h"
#include "TokenRequest.h"
#include "TokenResult.h"
```

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.h [hunk 14]
```diff
@@ -23,36 +23,23 @@
...
 
-#include "AuthenticationApi.h"
-#include "TokenEndpoint.h"
-#include "TokenRequest.h"
-#include "TokenResult.h"
+#include <olp/authentication/AuthenticationApi.h>
 
+#include "TokenEndpoint.h"
 #include "olp/core/client/CancellationContext.h"
 #include "olp/core/client/CancellationToken.h"
```
Comment (mykolamalik) [inline olp-cpp-sdk-authentication/src/AutoRefreshingToken.h:30]:
```suggestion
#include <olp/authentication/AuthenticationApi.h>
#include <olp/core/client/CancellationContext.h>
#include <olp/core/client/CancellationToken.h>
#include "TokenEndpoint.h"
```

File: olp-cpp-sdk-authentication/src/TokenProvider.cpp [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+#include <olp/authentication/TokenProvider.h>
+
+#include "TokenProviderImplPrivate.h"
+#include "TokenResult.h"
+
+namespace olp {
+namespace authentication {
+
+internal::TokenProviderImpl::TokenProviderImpl(
```
Comment (mykolamalik) [inline olp-cpp-sdk-authentication/src/TokenProvider.cpp:28]:
```suggestion
namespace authentication {
namespace internal{

TokenProviderImpl::TokenProviderImpl(
```

File: olp-cpp-sdk-authentication/src/TokenResult.h [hunk 16]
```diff
@@ -19,14 +19,12 @@
...
 #include <ctime>
 #include <string>
 
-#include "AuthenticationApi.h"
-#include "ErrorResponse.h"
+#include <olp/authentication/AuthenticationApi.h>
+#include <olp/authentication/ErrorResponse.h>
 
 namespace olp {
 namespace authentication {
```
Comment (mykolamalik) [inline olp-cpp-sdk-authentication/src/TokenResult.h:30]: Cannot add a suggestion for unchanged lines :(
class TokenResult must be without AUTHENTICATION_API

File: tests/integration/olp-cpp-sdk-authentication/TokenEndpointTest.cpp [hunk 17]
```diff
@@ -30,7 +30,10 @@
 #include <olp/core/http/NetworkConstants.h>
 #include <olp/core/http/NetworkUtils.h>
 #include <olp/core/porting/make_unique.h>
+#include <olp/core/porting/warning_disable.h>
```
Comment (mykolamalik) [inline tests/integration/olp-cpp-sdk-authentication/TokenEndpointTest.cpp:33]: Do we still need to disable warnings in this file, after most of the deprecations were removed?

**Review by mykolamalik — CHANGES_REQUESTED** [review]

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 1 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h]: done

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 2 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h]: done

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 3 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h]: ok

File: tests/integration/olp-cpp-sdk-authentication/TokenEndpointTest.cpp (hunk 17 above)
Comment (dudnyk) [reply tests/integration/olp-cpp-sdk-authentication/TokenEndpointTest.cpp:33]: I will enable warnings here)

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h [hunk 18]
```diff
@@ -134,79 +147,12 @@ class TokenProvider {
...
-    }
-
-   private:
-    std::chrono::seconds minimum_validity_{kDefaultMinimumValidity};
-    AutoRefreshingToken token_;
-    mutable std::mutex request_mutex_;
-  };
-
-  std::shared_ptr<TokenProviderImpl> impl_;
+  std::shared_ptr<olp::authentication::internal::TokenProviderImpl> impl_;
```
[the remark this answers is not in the document]
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:162]: done

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.h [hunk 19]
```diff
@@ -23,36 +23,24 @@
...
-
 /**
  * @brief Manages token requests.
  *
  * Requests a new token from the token endpoint and automatically refreshes it
  * when the token is about to expire.
  */
-class AUTHENTICATION_API OLP_SDK_DEPRECATED("Will be removed by 10.2020.")
-    AutoRefreshingToken {
+class AUTHENTICATION_API AutoRefreshingToken {
```
[the remark this answers is not in the document]
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/AutoRefreshingToken.h:41]: done

File: olp-cpp-sdk-authentication/src/TokenEndpoint.h [hunk 20]
```diff
@@ -26,30 +26,25 @@
...
 class TokenEndpointImpl;
+class TokenResult;
 
 /**
  * @brief Corresponds to the token endpoint as specified in the OAuth2.0
  * specification.
  */
-class AUTHENTICATION_API OLP_SDK_DEPRECATED("Will be removed in 10.2020")
-    TokenEndpoint {
+class AUTHENTICATION_API TokenEndpoint {
```
[the remark this answers is not in the document]
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenEndpoint.h:45]: done

File: olp-cpp-sdk-authentication/src/TokenRequest.h [hunk 21]
```diff
@@ -22,17 +22,15 @@
...
 
 namespace olp {
 namespace authentication {
 /**
  * @brief Holds the parameters of the OAuth2.0 Authorization
  * Grant request.
  */
-class AUTHENTICATION_API OLP_SDK_DEPRECATED("Will be removed in 10.2020")
-    TokenRequest {
+class AUTHENTICATION_API TokenRequest {
```
[the remark this answers is not in the document]
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenRequest.h:33]: done

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp (hunk 13 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/AutoRefreshingToken.cpp:29]: done

File: olp-cpp-sdk-authentication/src/AutoRefreshingToken.h (hunk 14 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/AutoRefreshingToken.h:30]: done

File: olp-cpp-sdk-authentication/src/TokenProvider.cpp (hunk 15 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenProvider.cpp:28]: done

File: olp-cpp-sdk-authentication/src/TokenResult.h (hunk 16 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenResult.h:30]: done

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h [hunk 22]
```diff
@@ -25,22 +25,47 @@
...
+static constexpr auto kDefaultMinimumValidity = 300ll;
+static constexpr auto kDefaultMinimumValiditySeconds =
+    std::chrono::seconds(kDefaultMinimumValidity);
+static constexpr auto kForceRefresh = std::chrono::seconds(0);
+
+namespace internal {
+
+class TokenProviderImplPrivate;
+
+// TokenProviderImpl should only be used as an implementation of TokenProvider
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:47]:
```suggestion
/// An implementation of `TokenProvider`.
```

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp [hunk 23]
```diff
@@ -0,0 +1,83 @@
...
+  const auto response = GetResponse(context);
+  return response ? response.GetResult().GetHttpStatus()
+                  : response.GetError().GetHttpStatusCode();
+}
+
+TokenResponse TokenProviderImplPrivate::GetResponse(
+    client::CancellationContext &context) const {
+  // Mutex is needed to prevent multiple authorization requests that can
+  // happen when the token is not available, and multiple consumers
+  // requested it.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp:69]:
```suggestion
  // Prevents multiple authorization requests that can
  // happen when the token is not available, and multiple consumers
  // request it.
```

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp [hunk 24]
```diff
@@ -0,0 +1,83 @@
...
+  std::lock_guard<std::mutex> lock(request_mutex_);
+  return token_->GetToken(context, minimum_validity_);
+}
+
+bool TokenProviderImplPrivate::IsTokenResponseOK() const {
+  client::CancellationContext context;
+
+  // Token response is successful if and only if we have a valid token.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp:77]:
```suggestion
  // The token response is successful only if the token is valid.
```

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+  client::OauthTokenResponse operator()(
+      client::CancellationContext& context) const;
+
+  ErrorResponse GetErrorResponse() const;
+
+  int GetHttpStatusCode() const;
+
+  TokenResponse GetResponse(client::CancellationContext& context) const;
+
+  bool IsTokenResponseOK() const;
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h:56]: All public members should be documented.

**Review by mykolamalik — APPROVED** [review]

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 22 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:47]: done

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp (hunk 23 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp:69]: done

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h (hunk 25 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h:56]: done

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp (hunk 24 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp:77]: done

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h [hunk 26]
```diff
@@ -25,22 +25,61 @@
...
+static constexpr auto kDefaultMinimumValidity = 300ll;
+static constexpr auto kDefaultMinimumValiditySeconds =
+    std::chrono::seconds(kDefaultMinimumValidity);
+static constexpr auto kForceRefresh = std::chrono::seconds(0);
+
+namespace internal {
+
+class TokenProviderImplPrivate;
+
+/// An implementation of `TokenProvider`.
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:47]: Please add something like:
```cpp
/// An implementation of `TokenProvider`.
/// @note This is a private implementation class for internal use only, and not bound
/// to any API stability promises. Please do not use directly.
class TokenProviderImpl {
 public:
```

File: olp-cpp-sdk-authentication/include/olp/authentication/Types.h [hunk 27]
```diff
@@ -59,6 +58,7 @@ using UserAccountInfoResponse = Response<model::UserAccountInfoResponse>;
 using UserAccountInfoCallback = Callback<model::UserAccountInfoResponse>;
 
 /// The access token response type.
+class TokenResult;
 using TokenResponse = Response<TokenResult>;
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/Types.h:62]: Moving the TokenResult to private makes this impossible to use for any outside library linked against Data SDK C++. Please remove.

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp [hunk 28]
```diff
@@ -0,0 +1,83 @@
+/*
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp:1]: I guess we can put this also in TokenProvider.cpp as we reached our goal of exposing internals trough the header.

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h [hunk 29]
```diff
@@ -0,0 +1,82 @@
+/*
```
Comment (andescu) [inline olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h:1]: Same, i think it is ok to put this in TokenProvider.cpp

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+      client::CancellationContext& context) const;
+
+  /// @copydoc TokenProvider::GetErrorResponse()
+  ErrorResponse GetErrorResponse() const;
+
+  /// @copydoc TokenProvider::GetHttpStatusCode()
+  int GetHttpStatusCode() const;
+
+  /// Gets the token response from `AutoRefreshingToken` or requests a new
+  /// token if the token response expired or is not present.
```
Comment (HalynaDumych) [inline olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h:68]:
```suggestion
  /**
   * @brief Gets the token response from `AutoRefreshingToken` or requests a new
   * token if the token response is expired or not present.
   *
   * @return The token response or the new token.
   */
```

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 26 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:47]: done

File: olp-cpp-sdk-authentication/include/olp/authentication/Types.h (hunk 27 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/Types.h]: removed, added header

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp (hunk 28 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.cpp]: Moved TokenProviderImplPrivate to TokenProvider.cpp

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h (hunk 29 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h]: Moved implementation of TokenProviderImplPrivate to TokenProvider.cpp

File: olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h (hunk 30 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/src/TokenProviderImplPrivate.h]: done(was moved to TokenProvider.cpp)

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h [hunk 31]
```diff
@@ -25,22 +25,63 @@
...
-PORTING_PUSH_WARNINGS()
-PORTING_CLANG_GCC_DISABLE_WARNING("-Wdeprecated-declarations")
+static constexpr auto kDefaultMinimumValidity = 300ll;
+static constexpr auto kDefaultMinimumValiditySeconds =
+    std::chrono::seconds(kDefaultMinimumValidity);
+static constexpr auto kForceRefresh = std::chrono::seconds(0);
+
+namespace internal {
+
+class TokenProviderImplPrivate;
```
Comment (andescu) [inline olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:45]:
```suggestion
class TokenProviderPrivate;
```

File: olp-cpp-sdk-authentication/src/TokenProvider.cpp (hunk 10 above)
Comment (andescu) [reply olp-cpp-sdk-authentication/src/TokenProvider.cpp:25]: I know what you mean, but namespace enclosing is a must.

File: olp-cpp-sdk-dataservice-write/tests/CMakeLists.txt [hunk 32]
```diff
@@ -107,6 +107,7 @@ else()
     target_include_directories(olp-cpp-sdk-dataservice-write-tests
     PRIVATE
         ${${PROJECT_NAME}_SOURCE_DIR}/src
+        ${PROJECT_SOURCE_DIR}/../olp-cpp-sdk-authentication/src
```
Comment (andescu) [inline olp-cpp-sdk-dataservice-write/tests/CMakeLists.txt:110]: Why is this needed? This usually indicates we have an API break?

File: tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp [hunk 33]
```diff
@@ -39,6 +37,7 @@
 #include <olp/core/porting/make_unique.h>
 
 #include "ApiClientLookup.h"
+#include "TokenEndpoint.h"
```
Comment (andescu) [inline tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp:40]: Why is TokenEndpoint needed? Can you please have a look and change the code if needed.

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp [hunk 34]
```diff
@@ -48,6 +46,7 @@
 #include <olp/dataservice/read/CatalogClient.h>
 #include <olp/dataservice/read/model/Catalog.h>
 #include <olp/dataservice/read/model/VersionInfos.h>
+#include "TokenEndpoint.h"
```
Comment (andescu) [inline tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp:49]: Same here.

File: tests/integration/olp-cpp-sdk-authentication/TokenEndpointTest.cpp [hunk 35]
```diff
@@ -17,7 +17,7 @@
  * License-Filename: LICENSE
  */
 
-#include <olp/authentication/TokenEndpoint.h>
+#include "TokenEndpoint.h"
```
Comment (andescu) [inline tests/integration/olp-cpp-sdk-authentication/TokenEndpointTest.cpp:20]: Move down to the list of project includes.
The rule of own header first is only for own header and this is not the header of TokenEndpointTest.cpp

File: olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h (hunk 31 above)
Comment (dudnyk) [reply olp-cpp-sdk-authentication/include/olp/authentication/TokenProvider.h:45]: done

File: olp-cpp-sdk-dataservice-write/tests/CMakeLists.txt (hunk 32 above)
Comment (dudnyk) [reply olp-cpp-sdk-dataservice-write/tests/CMakeLists.txt:110]: removed

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp (hunk 34 above)
Comment (dudnyk) [reply tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp:49]: removed

File: tests/integration/olp-cpp-sdk-authentication/TokenEndpointTest.cpp (hunk 35 above)
Comment (dudnyk) [reply tests/integration/olp-cpp-sdk-authentication/TokenEndpointTest.cpp:20]: got it

File: tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp (hunk 33 above)
Comment (dudnyk) [reply tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp:40]: removed

**Review by andescu — APPROVED** [review]

**Review by mykhailo-kuchma — APPROVED** [review]
